# [OLD] Integration site analysis on AAV vectors

- [ ] Check all these scripts for issues/bugs
- [ ] Add logs
- [ ] Remove input files

## Workflow

Here, a bunch of scripts to do ISA analysis on demultiplexed and error-corrected linked-read sequencing data generated by the TELL-Seq technology:
0. **TELL-Seq analysis**: run tell-seq pipeline and home made scripts to plot the results
1. **Clean mapping results**: filter out unmapped reads, duplicates reads, low mapping quality (MAPQ=0), similar reads in homologous regions between the two references and reads located in repeat regions
2. **Filter chimeric reads**: filter out reads mapped either on mouse or vector reference but not on both. Subset the initial mapping BAM file to keep only chimeric reads
3. **Reconstruct molecules**: adaptation from the original 10X Genomics' script to reconstruct molecules from barcoded reads.
4. **Identify chimeric molecules**: subset the BAM file to keep only molecules containing reads mapped to both references
5. **Identify potential IS**: extract information on chimeric molecules (name, chr, start, end) and identify non-redundant integration sites and regions of the vector involved in these integration events
6. **Structure analysis of potential IS**: blast each chimeric molecule identified to the two references to check reads involved in these integration events
7. **Chimeric molecule local assembly**: local assembly of chimeric molecules identified from the list of barcodes involved in these integration events


## Optional steps

There is also scripts about optional steps:
- **GC content**: use NTContent to have the distribution of GC content across the reference
- **Mark duplicates**: multiples methods used (clumpify, picard mark duplicates, samtools markdup)
- **Recode BAM header**: recode wrong chomosome coordinates order in the header of BAM mapping files with a home-made python script *recode_chr_name.py*
- **Rian locus local assembly**: isolate reads from the Rian locus and do local assembly with the Tell-Link module of the TELL-Seq software


## Visualization

There is also scripts about tools used for the visualization of the results:
- **chromoMap**: provide a configurable visualization of chromosomes
- **plotly**: coverage per position across the whole vector
- **karyoploteR**: density, coverage, distribution of reads across the whole mouse/vector genome
- **ggplot**: 
  - boxplot on number of molecules per barcode
  - histogram on number of reads per molecule OR length of molecules
